### PR TITLE
Fix #7002: Always invalidate virtual floor regardless of selection.

### LINKED ISSUE #7002: Issues with Virtual Floor invalidation
The Virtual Floor seems to have invalidation issues.

According to @JeroenDStout it performs better in terms of displaying correctly at higher frame rates.

image

### PR DESCRIPTION
**Pull request by ZehMatt**
Currently the virtual floor would not invalidate if placement would be impossible, this should correct the issue.

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
Thanks, this appears to fix the horizontal invalidation problems nicely. It does not fix the redrawing lag, but that's a different issue, of course.

**Comment by ZehMatt** [comment]
What would be the best way to reproduce the redraw lag?

**Comment by AaronVanGeffen** [comment]
Move the selection up and down quickly. You'll notice part of the floor is lagging behind. For me, it's usually the centre not being redrawn, while the outer parts are. It's only there for a fraction of a second, but noticeable nonetheless.

**Comment by ZehMatt** [comment]
I can only get it slightly to lag with software rendering not with opengl, could you please test that on your end?

**Comment by AaronVanGeffen** [comment]
Yeah, it only happens with the software renderer. OpenGL does not have the problem, as it doesn't use invalidation.

**Comment by ZehMatt** [comment]
Makes sense, at least I know where to look now, thank you.

**Comment by ZehMatt** [comment]
This should do. Would be nice if someone can confirm it.

File: src/openrct2/world/Map.cpp [hunk 1]
```diff
@@ -4819,11 +4819,6 @@ void map_invalidate_virtual_floor_tiles()
         }
     }
 
-    // Do not invalidate if we're between ticks.
-    if (min_position.x == std::numeric_limits<sint16>::max()    || min_position.y == std::numeric_limits<sint16>::max() ||
-        max_position.x == std::numeric_limits<sint16>::lowest() || max_position.y == std::numeric_limits<sint16>::lowest())
```
Comment (AaronVanGeffen) [inline src/openrct2/world/Map.cpp]: This was some very deliberate coding — removing it has a very real performance impact. Without it, FPS drops dramatically with the virtual floor active: down to 10fps (active) from 70fps (inactive) here.

Comment (ZehMatt) [reply src/openrct2/world/Map.cpp]: But does it actually solve the issue? I'll have a look at the performance issue, I haven't seen it on my end.

Comment (AaronVanGeffen) [reply src/openrct2/world/Map.cpp]: Yes, removing it does solve the issue with rendering… If you could achieve that without the dip in FPS, I'll happily merge.

**Comment by ZehMatt** [comment]
I've moved most of the virtual floor code to VirtualFloor.cpp and changed the way it invalidates the tiles. I will clean up the code once I know all of it is working properly now.

**Comment by AaronVanGeffen** [comment]
You've hit the nail on the head this time — it's faster than it's ever been, for me, too. Cheers. If you could clean up the transformation functions, I'll be happy to merge this. (You can make the originals non-static and export them through Map.h, if you like.)

File: src/openrct2/paint/VirtualFloor.cpp [hunk 2]
```diff
@@ -30,6 +45,181 @@ static constexpr const CoordsXY offsets[4] =
...
+    _virtualFloorLastMinPos.x = std::numeric_limits<sint16>::max();
+    _virtualFloorLastMinPos.y = std::numeric_limits<sint16>::max();
+    _virtualFloorLastMaxPos.x = std::numeric_limits<sint16>::lowest();
+    _virtualFloorLastMaxPos.y = std::numeric_limits<sint16>::lowest();
+    _virtualFloorHeight = 0;
+}
+
+void virtual_floor_enable()
+{
+    if (virtual_floor_is_enabled() == true)
```
Comment (AaronVanGeffen) [inline src/openrct2/paint/VirtualFloor.cpp]: Now returns a bool, so you can leave out the ` == true`.

File: src/openrct2/paint/VirtualFloor.cpp [hunk 3]
```diff
@@ -30,6 +45,181 @@ static constexpr const CoordsXY offsets[4] =
...
+        return;
+    }
+
+    virtual_floor_reset();
+    _virtualFloorFlags |= VIRTUAL_FLOOR_FLAG_ENABLED;
+}
+
+void virtual_floor_disable()
+{
+    if (virtual_floor_is_enabled() == false)
```
Comment (AaronVanGeffen) [inline src/openrct2/paint/VirtualFloor.cpp]: Now returns a bool, so you can leave out the ` == false`.

**Pull request merged**